docs/architecture/adr-008-ics20-implementation.md

```diff
@@ -0,0 +1,298 @@
+    pub struct Coin {
+        /// Denomination
+        pub denom: Denom,
+
+        /// Amount
+        pub amount: Decimal,
+    }
+
+    impl TryFrom<proto::cosmos::base::v1beta1::Coin> for Coin {
+        type Error = ICS20Error;
+
+        fn try_from(proto: proto::cosmos::base::v1beta1::Coin) -> Result<Coin, Self::Error> {
+            ...
+        }
+    }
+
+    impl TryFrom<&proto::cosmos::base::v1beta1::Coin> for Coin {
+        type Error = ICS20Error;
+
+        fn try_from(proto: &proto::cosmos::base::v1beta1::Coin) -> Result<Coin, Self::Error> {
+            ...
+        }
+    }
+
+    impl From<Coin> for proto::cosmos::base::v1beta1::Coin {
+        fn from(coin: Coin) -> proto::cosmos::base::v1beta1::Coin {
+            ...
+        }
+    }
+
+    impl From<&Coin> for proto::cosmos::base::v1beta1::Coin {
+        fn from(coin: &Coin) -> proto::cosmos::base::v1beta1::Coin {
+            ...
+        }
+    }
+
+    pub struct TracePrefix {
+        pub port_id: PortId,
+        pub channel_id: ChannelId
+    }
+
+    /// This struct and it's implementations should help identifying denomination traces
+    pub struct DenomTrace {
+        /// Full denomination path
+        pub trace_path: Vec<TracePrefix>,
+        /// Base denimination
+        pub base_denom: Denom
+    }
+
+    impl DenomTrace {
+        /// Returns the full denom path
+        fn get_full_denom_path(&self) -> String {
+            ...
+        }
+        /// IBCDenom a coin denomination for an ICS20 fungible token in the format
+        /// 'ibc/trace_path/base_denom'. If the trace is empty, it will return the base denomination.
+        fn ibc_denom(&self) -> String {
+            ...
+        }
+        /// Returns the prefix for this trace
+        fn get_prefix(&self) -> String {
+            ...
+        }
+
+        /// Returns the prefix for this trace
+        fn has_prefix(full_path: &str, prefix: &str) -> bool {
+            ...
+        }
+    }
+
+    impl FromStr for DenomTrace {
+        ...
+    }
+
+    pub trait ICS20Keeper: 
+        ChannelKeeper 
+        + PortKeeper 
+        + PortReader 
+        + BankKeeper<Self::AccountId> 
+        + AccountKeeper<Self::AccountId> 
+        + CapabilityKeeper 
+    {
+        type AccountId: Into<String>;
+        /// bind_port defines a wrapper function for the PortKeeper's bind_port function.
+        fn bind_port(&self, port_id: PortId) -> Result<(), ICS20Error>;
+        /// set_port sets the portID for the transfer module.
+        fn set_port(&self, port_id: PortId) -> ();
+        /// authenticate_capability wraps the CapabilityKeeper's authenticate_capability function
+        fn authenticate_capability(&self, cap: Capability, name: &str) -> bool;
+        /// claim_capability allows the transfer module to claim a capability that IBC module
+        /// passes to it
+        fn claim_capability(&self, cap: Capability, name: &str) -> Result<(), ICS20Error>;
```

Comment:
I think this keeper must provide a function to modify the `channelEscrowAddresses` mapping in the module state.